### ISSUE TITLE
Lint for guarded unreachable match expression patterns

### ISSUE THREAD

**Issue by edward-shen** [rust-lang/rust-clippy#5100]
Clippy version: `0.0.212 (c8e3cfb 2019-10-28)`

Clippy will detect unreachable patterns if they're not guarded against:

```rust
enum Foo {
    Hello,
    World,
}

fn main() {
    let foo = Foo::Hello;
    
    match foo {
        Foo::Hello | Foo::World => println!("hello world!"),
        Foo::Hello => println!("this should be dead code"),
        Foo::World => println!("this too should be dead code"),
    }
}
```
```
warning: unreachable pattern
  --> src/main.rs:19:9
   |
19 |         Foo::Hello => println!("this should be dead code"),
   |         ^^^^^^^^^^
   |
   = note: `#[warn(unreachable_patterns)]` on by default

warning: unreachable pattern
  --> src/main.rs:20:9
   |
20 |         Foo::World => println!("this too should be dead code"),
   |         ^^^^^^^^^^
```

However, clippy fails to do so if the arm is guarded against:

```rust
enum Foo {
    Hello,
    World,
}

fn main() {
    let foo = Foo::Hello;
    let bar = true;
    
    match foo {
        Foo::Hello | Foo::World if bar => println!("hello world!"),
        Foo::Hello if bar => println!("this should be dead code"),
        Foo::World if bar => println!("this too should be dead code"),
        _ => println!("goodbye world!"),
    }
}
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d68a0fd092b8c1f4341151de20365fa5)

On another note, changing foo to `Foo::World` will warn that `Foo::Hello` is never constructed, but fails to do so the other way around—should I open another issue for that?

**Comment by JohnTitor** [comment]
Note that `unreachable_patterns` is not a clippy lint but a rustc lint.

**Comment by flip1995** [comment]
This is a `rustc` (compiler) lint, not a Clippy lint. 

It's hard to improve this behavior, since the guard can be both `true` or `false` and can rarely be const evaluated to one of them. So in your example, every pattern is reachable if `bar` is `false`.

**Comment by edward-shen** [comment]
Ah, I see. Thanks for taking a look!
